docs/source/notes/distributed_autograd.rst

```diff
@@ -0,0 +1,278 @@
+Distributed Autograd Design
+===========================
+
+This note will present the detailed design for distributed autograd and walk 
+through the internals of the same. Make sure you're familiar with 
+`Autograd Mechanics`_, 
+``torch.distributed.rpc`` and ``RRef`` before proceeding 
+(RFC: https://github.com/pytorch/pytorch/issues/23110).
+
+.. 
+  TODO: Update the link above to point to actual RPC and RRef docs.
+
+.. background: 
+
+Background
+^^^^^^^^^^
+
+Let's say you have two nodes and a very simple model partitioned across two 
+nodes. This can be implemented using ``torch.distributed.rpc`` as follows:
+
+.. code::
+
+  import torch
+  import torch.distributed.rpc as rpc
+
+  def my_add(t1, t2):
+    return torch.add(t1, t2)
+
+  # On worker 0:
+  t1 = torch.rand((3, 3), requires_grad=True)
+  t2 = torch.rand((3, 3), requires_grad=True)
+
+  # Perform some computation remotely.
+  t3 = rpc.rpc_sync("worker1", my_add, args=(t1, t2))
+
+  # Perform some computation locally based on remote result.
+  t4 = torch.rand((3, 3), requires_grad=True)
+  t5 = torch.mul(t3, t4)
+
+  # Compute some loss.
+  loss = t5.sum()
+
+The main motivation behind distributed autograd is to enable running a backward 
+pass on such distributed models with the ``loss`` that we've computed and 
+record appropriate gradients for all tensors that require gradients.
+
+.. attaching_send_recv_functions:
+
+Autograd recording during the forward pass
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+PyTorch builds the autograd graph during the forward pass and this graph is 
+used to execute the backward pass. For more details see 
+`How autograd encodes history`_. 
+
+For distributed autograd, we need to keep track of all RPCs during the forward 
+pass to ensure the backward pass is executed appropriately. For this purpose, 
+we attach ``send`` and ``recv`` functions to the autograd graph when we perform 
+an RPC.
+
+- The ``send`` function is attached to the source of the RPC and it's output 
+  edges point to the autograd function for the input tensors of the RPC. 
+  The input for this function is received over the wire from the appropriate 
```

Comment:
Suggestion : The phrase "received over the wire" is very American idiomatic. Can be challenging for non-native speakers to understand/difficult to translate. Consider simplifying to just "is received from the appropriate..." or something similar.